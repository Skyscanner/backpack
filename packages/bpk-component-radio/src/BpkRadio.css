@keyframes bpk-keyframe-spin{100%{transform:rotate(1turn)}}.bpk-radio{position:relative;display:inline-block;padding-left:1.5rem}html[dir=rtl] .bpk-radio{padding-right:1.5rem;padding-left:0}.bpk-radio--white{color:#fff}.bpk-radio--disabled{color:#b2b2bf;cursor:not-allowed}.bpk-radio--invalid input[type=radio]{background:url("data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAxOCAxOCI+PGcgZmlsbD0ibm9uZSIgZmlsbC1ydWxlPSJldmVub2RkIj48cGF0aCBmaWxsPSIjRkZGIiBkPSJNOSAwYTkgOSAwIDAgMSA5IDkgOSA5IDAgMCAxLTkgOSA5IDkgMCAwIDEtOS05IDkgOSAwIDAgMSA5LTl6Ii8+PHBhdGggZmlsbD0iI0QxNDM1QiIgZD0iTTkgMTdhOCA4IDAgMCAwIDgtOCA4IDggMCAwIDAtOC04IDggOCAwIDAgMC04IDggOCA4IDAgMCAwIDggOHpNOSAwYTkgOSAwIDAgMSA5IDkgOSA5IDAgMCAxLTkgOSA5IDkgMCAwIDEtOS05IDkgOSAwIDAgMSA5LTl6Ii8+PC9nPjwvc3ZnPg==") no-repeat}.bpk-radio__input{position:absolute;top:.1875rem;left:0;width:1.125rem;height:1.125rem;margin:0;padding:0;border:0;background:url("data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAxOCAxOCI+PGcgZmlsbD0ibm9uZSIgZmlsbC1ydWxlPSJldmVub2RkIj48cGF0aCBmaWxsPSIjRkZGIiBkPSJNOSAwYTkgOSAwIDAgMSA5IDkgOSA5IDAgMCAxLTkgOSA5IDkgMCAwIDEtOS05IDkgOSAwIDAgMSA5LTl6Ii8+PHBhdGggZmlsbD0iI0IyQjJCRiIgZD0iTTkgMTdhOCA4IDAgMCAwIDgtOCA4IDggMCAwIDAtOC04IDggOCAwIDAgMC04IDggOCA4IDAgMCAwIDggOHpNOSAwYTkgOSAwIDAgMSA5IDkgOSA5IDAgMCAxLTkgOSA5IDkgMCAwIDEtOS05IDkgOSAwIDAgMSA5LTl6Ii8+PC9nPjwvc3ZnPg==") no-repeat;cursor:pointer;vertical-align:text-bottom;appearance:none}html[dir=rtl] .bpk-radio__input{right:0;left:auto}@media screen\0 {.bpk-radio__input{background:none !important}}.bpk-radio__input:checked{background:url("data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAxOCAxOCI+PGcgZmlsbD0ibm9uZSIgZmlsbC1ydWxlPSJldmVub2RkIj48cGF0aCBmaWxsPSIjRkZGIiBkPSJNOSAwYTkgOSAwIDAgMSA5IDkgOSA5IDAgMCAxLTkgOSA5IDkgMCAwIDEtOS05IDkgOSAwIDAgMSA5LTl6Ii8+PHBhdGggZmlsbD0iI0IyQjJCRiIgZD0iTTkgMTdhOCA4IDAgMCAwIDgtOCA4IDggMCAwIDAtOC04IDggOCAwIDAgMC04IDggOCA4IDAgMCAwIDggOHpNOSAwYTkgOSAwIDAgMSA5IDkgOSA5IDAgMCAxLTkgOSA5IDkgMCAwIDEtOS05IDkgOSAwIDAgMSA5LTl6Ii8+PGNpcmNsZSBjeD0iOSIgY3k9IjkiIHI9IjUiIGZpbGw9IiMwMDhjYTgiLz48L2c+PC9zdmc+") no-repeat}.bpk-radio__input:checked:disabled{background:url("data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAxOCAxOCI+PGcgZmlsbD0iI0IyQjJCRiI+PHBhdGggZD0iTTkgMGM0Ljk3MSAwIDkgNC4wMjggOSA5IDAgNC45NzEtNC4wMjggOS05IDktNC45NzEgMC05LTQuMDI4LTktOSAwLTQuOTcxIDQuMDI4LTkgOS05eiIvPjxwYXRoIGQ9Ik05IDE3YTggOCAwIDAgMCA4LTggOCA4IDAgMCAwLTgtOCA4IDggMCAwIDAtOCA4IDggOCAwIDAgMCA4IDh6TTkgMGM0Ljk3MSAwIDkgNC4wMjggOSA5IDAgNC45NzEtNC4wMjggOS05IDktNC45NzEgMC05LTQuMDI4LTktOSAwLTQuOTcxIDQuMDI4LTkgOS05eiIvPjxjaXJjbGUgY3g9IjkiIGN5PSI5IiByPSI1Ii8+PC9nPjwvc3ZnPg==") no-repeat}.bpk-radio__input:disabled{background:url("data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAxOCAxOCI+PGcgZmlsbD0iI0IyQjJCRiI+PHBhdGggZD0iTTkgMGM0Ljk3MSAwIDkgNC4wMjggOSA5IDAgNC45NzEtNC4wMjggOS05IDktNC45NzEgMC05LTQuMDI4LTktOSAwLTQuOTcxIDQuMDI4LTkgOS05eiIvPjxwYXRoIGQ9Ik05IDE3YTggOCAwIDAgMCA4LTggOCA4IDAgMCAwLTgtOCA4IDggMCAwIDAtOCA4IDggOCAwIDAgMCA4IDh6TTkgMGM0Ljk3MSAwIDkgNC4wMjggOSA5IDAgNC45NzEtNC4wMjggOS05IDktNC45NzEgMC05LTQuMDI4LTktOSAwLTQuOTcxIDQuMDI4LTkgOS05eiIvPjwvZz48L3N2Zz4=") no-repeat}.bpk-radio__input:checked{background:url("data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAxOCAxOCI+PGcgZmlsbD0ibm9uZSIgZmlsbC1ydWxlPSJldmVub2RkIj48cGF0aCBmaWxsPSIjRkZGIiBkPSJNOSAwYTkgOSAwIDAgMSA5IDkgOSA5IDAgMCAxLTkgOSA5IDkgMCAwIDEtOS05IDkgOSAwIDAgMSA5LTl6Ii8+PHBhdGggZmlsbD0iI0IyQjJCRiIgZD0iTTkgMTdhOCA4IDAgMCAwIDgtOCA4IDggMCAwIDAtOC04IDggOCAwIDAgMC04IDggOCA4IDAgMCAwIDggOHpNOSAwYTkgOSAwIDAgMSA5IDkgOSA5IDAgMCAxLTkgOSA5IDkgMCAwIDEtOS05IDkgOSAwIDAgMSA5LTl6Ii8+PC9nPjwvc3ZnPg==") no-repeat}.bpk-radio__input:checked+.bpk-radio__circle{display:block}.bpk-radio__input:checked:disabled{background:url("data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAxOCAxOCI+PGcgZmlsbD0iI0IyQjJCRiI+PHBhdGggZD0iTTkgMGM0Ljk3MSAwIDkgNC4wMjggOSA5IDAgNC45NzEtNC4wMjggOS05IDktNC45NzEgMC05LTQuMDI4LTktOSAwLTQuOTcxIDQuMDI4LTkgOS05eiIvPjxwYXRoIGQ9Ik05IDE3YTggOCAwIDAgMCA4LTggOCA4IDAgMCAwLTgtOCA4IDggMCAwIDAtOCA4IDggOCAwIDAgMCA4IDh6TTkgMGM0Ljk3MSAwIDkgNC4wMjggOSA5IDAgNC45NzEtNC4wMjggOS05IDktNC45NzEgMC05LTQuMDI4LTktOSAwLTQuOTcxIDQuMDI4LTkgOS05eiIvPjxjaXJjbGUgY3g9IjkiIGN5PSI5IiByPSI1Ii8+PC9nPjwvc3ZnPg==") no-repeat}.bpk-radio__input:checked:disabled+.bpk-radio__circle{display:none}.bpk-radio__circle{position:absolute;top:.4375rem;left:.25rem;display:none;width:.625rem;height:.625rem;border-radius:50%;background:#0770e3;background:var(--bpk-radio-checked-color, #0770e3)}html[dir=rtl] .bpk-radio__circle{right:.25rem;left:auto}
