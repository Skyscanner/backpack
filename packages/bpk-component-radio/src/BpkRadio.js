/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow strict */

import PropTypes from 'prop-types';
import React, { type Node } from 'react';

import { cssModules } from '../../bpk-react-utils';

import STYLES from './BpkRadio.module.scss';

const getClassName = cssModules(STYLES);

type Props = {
  name: string,
  label: Node,
  ariaLabel: ?string,
  className: ?string,
  disabled: boolean,
  white: boolean,
  valid: ?boolean,
};

const BpkRadio = (props: Props) => {
  const { ariaLabel, className, disabled, label, name, valid, white, ...rest } =
    props;

  // Explicit check for false primitive value as undefined is
  // treated as neither valid nor invalid
  const isInvalid = valid === false;

  const classNames = getClassName(
    'bpk-radio',
    white && 'bpk-radio--white',
    disabled && 'bpk-radio--disabled',
    white && disabled && 'bpk-radio--disabled--white',
    isInvalid && 'bpk-radio--invalid',
    className,
  );

  return (
    <label className={classNames}>
      {/* Deciding to support this because `aria-invalid` does often work with voiceover
      despite not being in the spec. */}
      {/* eslint-disable jsx-a11y/role-supports-aria-props */}
      {/* $FlowFixMe[cannot-spread-inexact] - inexact rest. See decisions/flowfixme.md */}
      <input
        type="radio"
        className={getClassName('bpk-radio__input')}
        name={name}
        disabled={disabled}
        aria-label={ariaLabel || label}
        aria-invalid={isInvalid}
        {...rest}
      />
      {/* eslint-enable */}
      <div className={getClassName('bpk-radio__circle')} />
      <span aria-hidden="true">{label}</span>
    </label>
  );
};

BpkRadio.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.node.isRequired,
  ariaLabel: PropTypes.string,
  disabled: PropTypes.bool,
  white: PropTypes.bool,
  className: PropTypes.string,
  valid: PropTypes.bool,
};

BpkRadio.defaultProps = {
  ariaLabel: null,
  disabled: false,
  white: false,
  className: null,
  valid: null,
};

export default BpkRadio;
