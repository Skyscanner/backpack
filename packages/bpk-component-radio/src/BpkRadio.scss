/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016-2020 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

@import '~bpk-mixins/index';

$bpk-radio-circle-size: $bpk-one-pixel-rem * 10;

.bpk-radio {
  @include bpk-radio;

  &--white {
    @include bpk-radio--white;
  }

  &--disabled {
    @include bpk-radio--disabled;
  }

  &--invalid {
    @include bpk-radio--invalid;
  }

  &__input {
    @include bpk-radio__input;

    &:checked {
      background: url(map-get($bpk-elements, radio-normal)) no-repeat;

      + .bpk-radio__circle {
        display: block;
      }

      &:disabled {
        background: url(map-get($bpk-elements, radio-checked-disabled)) no-repeat;

        + .bpk-radio__circle {
          display: none;
        }
      }
    }
  }

  &__circle {
    position: absolute;
    top: ($bpk-spacing-base - $bpk-radio-circle-size) / 2;
    left: ($bpk-spacing-md - $bpk-radio-circle-size) / 2;
    display: none;
    width: $bpk-radio-circle-size;
    height: $bpk-radio-circle-size;
    border-radius: 50%;

    @include bpk-themeable-property(background, --bpk-radio-checked-color, $bpk-radio-checked-circle-color);

    @include bpk-rtl {
      right: ($bpk-spacing-md - $bpk-radio-circle-size) / 2;
      left: auto;
    }
  }
}
