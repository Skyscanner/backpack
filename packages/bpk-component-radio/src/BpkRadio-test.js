/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016-2019 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import renderer from 'react-test-renderer';

import BpkRadio from './BpkRadio';

describe('BpkRadio', () => {
  it('should render correctly', () => {
    const tree = renderer
      .create(<BpkRadio name="radio" label="Direct" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly when label is a node', () => {
    const tree = renderer
      .create(<BpkRadio name="radio" label={<span>Direct</span>} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with id attribute', () => {
    const tree = renderer
      .create(<BpkRadio id="radio" name="radio" label="Direct" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with checked attribute', () => {
    const tree = renderer
      .create(<BpkRadio name="radio" label="Direct" checked />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with disabled attribute', () => {
    const tree = renderer
      .create(<BpkRadio name="radio" label="Direct" disabled />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with white attribute', () => {
    const tree = renderer
      .create(<BpkRadio name="radio" label="Direct" white />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with value attribute', () => {
    const tree = renderer
      .create(<BpkRadio name="radio" label="Direct" value="my-value" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with ariaLabel attribute', () => {
    const tree = renderer
      .create(
        <BpkRadio
          id="radio"
          name="radio"
          label="Direct"
          ariaLabel="Is this direct"
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
