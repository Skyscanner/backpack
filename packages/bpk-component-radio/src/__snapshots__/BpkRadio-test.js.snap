// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BpkRadio should render correctly 1`] = `
<label
  className="bpk-radio"
>
  <input
    aria-label="Direct"
    className="bpk-radio__input"
    disabled={false}
    name="radio"
    type="radio"
  />
  <div
    className="bpk-radio__circle"
  />
  <span
    aria-hidden="true"
  >
    Direct
  </span>
</label>
`;

exports[`BpkRadio should render correctly when label is a node 1`] = `
<label
  className="bpk-radio"
>
  <input
    aria-label={
      <span>
        Direct
      </span>
    }
    className="bpk-radio__input"
    disabled={false}
    name="radio"
    type="radio"
  />
  <div
    className="bpk-radio__circle"
  />
  <span
    aria-hidden="true"
  >
    <span>
      Direct
    </span>
  </span>
</label>
`;

exports[`BpkRadio should render correctly with ariaLabel attribute 1`] = `
<label
  className="bpk-radio"
>
  <input
    aria-label="Is this direct"
    className="bpk-radio__input"
    disabled={false}
    id="radio"
    name="radio"
    type="radio"
  />
  <div
    className="bpk-radio__circle"
  />
  <span
    aria-hidden="true"
  >
    Direct
  </span>
</label>
`;

exports[`BpkRadio should render correctly with checked attribute 1`] = `
<label
  className="bpk-radio"
>
  <input
    aria-label="Direct"
    checked={true}
    className="bpk-radio__input"
    disabled={false}
    name="radio"
    type="radio"
  />
  <div
    className="bpk-radio__circle"
  />
  <span
    aria-hidden="true"
  >
    Direct
  </span>
</label>
`;

exports[`BpkRadio should render correctly with disabled attribute 1`] = `
<label
  className="bpk-radio bpk-radio--disabled"
>
  <input
    aria-label="Direct"
    className="bpk-radio__input"
    disabled={true}
    name="radio"
    type="radio"
  />
  <div
    className="bpk-radio__circle"
  />
  <span
    aria-hidden="true"
  >
    Direct
  </span>
</label>
`;

exports[`BpkRadio should render correctly with id attribute 1`] = `
<label
  className="bpk-radio"
>
  <input
    aria-label="Direct"
    className="bpk-radio__input"
    disabled={false}
    id="radio"
    name="radio"
    type="radio"
  />
  <div
    className="bpk-radio__circle"
  />
  <span
    aria-hidden="true"
  >
    Direct
  </span>
</label>
`;

exports[`BpkRadio should render correctly with value attribute 1`] = `
<label
  className="bpk-radio"
>
  <input
    aria-label="Direct"
    className="bpk-radio__input"
    disabled={false}
    name="radio"
    type="radio"
    value="my-value"
  />
  <div
    className="bpk-radio__circle"
  />
  <span
    aria-hidden="true"
  >
    Direct
  </span>
</label>
`;

exports[`BpkRadio should render correctly with white attribute 1`] = `
<label
  className="bpk-radio bpk-radio--white"
>
  <input
    aria-label="Direct"
    className="bpk-radio__input"
    disabled={false}
    name="radio"
    type="radio"
  />
  <div
    className="bpk-radio__circle"
  />
  <span
    aria-hidden="true"
  >
    Direct
  </span>
</label>
`;
